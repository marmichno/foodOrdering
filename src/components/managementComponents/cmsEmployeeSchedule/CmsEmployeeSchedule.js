import {CmsNavbar} from '../cmsNavbar/CmsNavbar';
import {useState, useEffect} from 'react';
import {employeesShiftPostRequest} from '../../../requests/employeesShiftPostRequest';
import {employeesShiftGetRequest} from '../../../requests/employeesShiftGetRequest';
import {employeesShiftDeleteRequest} from '../../../requests/employeesShiftDeleteRequest';
import {employeesSchedulePostRequest} from '../../../requests/employeesSchedulePostRequest';
import {employeesScheduleGetRequest} from '../../../requests/employeesScheduleGetRequest';
import {employeesScheduleDeleteRequest} from '../../../requests/employeesScheduleDeleteRequest';
import {employeesGetRequest} from '../../../requests/employeesGetRequest';

export const CmsEmployeesSchedule = () =>{

    const newDate = new Date();

    const [selectedMonth, setSelectedMonth] = useState(newDate.getMonth());
    const [selectedYear, setSelectedYear] = useState(newDate.getFullYear());
    const [shiftRangeText, setShiftRangeText] = useState("");
    const [deleteShiftId, setDeleteShiftId] = useState("");
    const [shifts, setShifts] = useState([]);
    const [selectedShift, setSelectedShift] = useState("");
    const [employees, setEmployees] = useState([]);
    const [schedule, setSchedule] = useState([]);

    useEffect(() => {
        getShifts();
        getEmployees();
        getSchedule();
    },[]);

    useEffect(() =>{
        getSchedule();
    },[selectedMonth]);


    const getEmployees = async () =>{
        const response = await employeesGetRequest();
        setEmployees(response);
    }
    
    const getShifts = async () =>{
        const response = await employeesShiftGetRequest();
        setShifts(response);
    }

    const addShift = async () =>{
        const response = await employeesShiftPostRequest({
            "range":shiftRangeText
        });
        getShifts();
    }

    const getSchedule = async () =>{
        const response = await employeesScheduleGetRequest(selectedYear, selectedMonth);
        setSchedule(response);
    }
    
    const deleteShift = async () =>{
        const response = await employeesShiftDeleteRequest(deleteShiftId);
        getShifts();
    }


    const monthNames = ["January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December"];

    let month = newDate.getMonth();

    const createArrayOfDays = (selectedMonth, selectedYear) => {
        let arrayOfDays = [];

        let daysInMonth = new Date(selectedYear, selectedMonth, 0).getDate();

        const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 
        'Thursday', 'Friday', 'Saturday'];

        for(let i = 0; i < daysInMonth; i++){
            let whichDayIsIt = new Date(selectedYear, selectedMonth, i + 1).getDay();

            arrayOfDays.push(
                {
                    dayDate: (i + 1) < 10 ? `0${i + 1}` : (i + 1),
                    selectedMonth: (selectedMonth + 1) < 10 ? `0${selectedMonth + 1}` : (selectedMonth + 1),
                    selectedYear: selectedYear,
                    dayName: days[whichDayIsIt]
                }
            );
        }

        return arrayOfDays;
    }

    const modifyEmployeeShift = async (e) => {

        const date  = e.currentTarget.dataset.date;
        const employeeEntity = e.currentTarget.dataset.employee;

        if(parseInt(selectedShift) !== 0 && selectedShift !== ""){
            const response = await employeesSchedulePostRequest(date, selectedShift, employeeEntity);
            getSchedule();
        }else if(parseInt(selectedShift) === 0){
            const shiftId = findShift(date, parseInt(employeeEntity), "deleteShift");

            if(shiftId === undefined){
                return;
            }else{
                const response = await employeesScheduleDeleteRequest(shiftId.id);
                getSchedule();
            }
        }else{
            return;
        }
    }

    const changeDate = (e) => {
        if(e.currentTarget.innerHTML === "next"){
            if(selectedMonth <= 10){
                setSelectedMonth(selectedMonth + 1);
            }else{
                setSelectedYear(selectedYear + 1);
                setSelectedMonth(0);
            }
        }else{
            if(selectedMonth !== 0){
                setSelectedMonth(selectedMonth - 1);
            }else{
                setSelectedMonth(11);
                setSelectedYear(selectedYear - 1);
            }
        }
    }

    const findShift = (shiftDate, shiftEmployeeId, event) =>{
    
        if(schedule !== undefined){
            const shift = schedule.filter(value => {
                    if(value.date === shiftDate && value.employeeEntity.id === shiftEmployeeId){
                        return true;
                    }else{
                        return false;
                    }
                })[0];

                if(event === "findShift"){
                    return shift !== undefined ? shift.shiftEntity.id : "";
                }else if(event === "deleteShift"){
                    return shift;
                }
        }else{
            return
        }
    }

    const changeSelectedShift = (e) =>{
        let shifts = document.querySelectorAll('.shiftActive');
        shifts.forEach(element => element.classList.remove('shiftActive'));
        e.currentTarget.classList.add('shiftActive');
        setSelectedShift(e.currentTarget.dataset.shift);
    }

    return(
        <div className="cmsEmployeesScheduleContainer">
            <CmsNavbar/>
            <div className="cmsEmployeesScheduleContent">
                <div className="shiftsContainer">
                    <div className="shiftsInfoContainer">
                        <div className="shiftInfo">
                            <input onChange={(e) => setDeleteShiftId(e.target.value)} type="number" placeholder="shift id" autocomplete="off" name="category" required></input>
                            <button onClick={deleteShift}>delete</button>
                        </div>
                        <div className="shiftInfo">
                            <h1 data-shift={0} onClick={changeSelectedShift}>day off</h1>
                        </div>
                        {shifts !== undefined ?
                            shifts.map(value => {
                                return <div className="shiftInfo">
                                            <h1 data-shift={value.id} onClick={changeSelectedShift}>
                                                {value.id} - {value.range}
                                            </h1>
                                        </div>
                            })
                        : null
                        }
                    </div>
                    <div className="addShifts">
                        <input onChange={(e) => setShiftRangeText(e.target.value)} type="text" placeholder="hour range ex: 16:30-24:30" autocomplete="off" name="category" required></input>
                        <button onClick={addShift}>add new shift</button>
                    </div>
                </div>

                <div className="calendarContainer">
                    
                    <div className="changeDate">
                        <button onClick={changeDate}>previous</button><h1>{`${selectedYear}, ${monthNames[selectedMonth]}`}</h1><button onClick={changeDate}>next</button>
                    </div>

                    <div className="employeeRow">
                    <div className="employee"></div>
                    {createArrayOfDays(selectedMonth, selectedYear).map(value =>{
                        return <div><b>{value.dayDate}</b></div>
                    })}
                    </div>
                    {
                    employees !== undefined ?
                        employees.map((employee, id) => {
                            return(
                                <>
                                    <div className="employeeRow">
                                        <div className="employee">{employee.firstName}</div>
                                            {createArrayOfDays(selectedMonth, selectedYear).map(value => {
                                                const date = `${value.selectedYear}-${value.selectedMonth}-${value.dayDate}`

                                                if(value.dayName === 'Saturday' || value.dayName === 'Sunday'){

                                                    return <div 
                                                    onClick={modifyEmployeeShift}
                                                    data-date={date}
                                                    data-employee={employee.id} 
                                                    style={{backgroundColor: 'orange'}}
                                                    >
                                                        {findShift(`${date}`, employee.id, "findShift")}
                                                    </div>
                                                }else{
                                                    return <div
                                                    onClick={modifyEmployeeShift}
                                                    data-date={date}
                                                    data-employee={employee.id}
                                                    >
                                                        {findShift(`${date}`, employee.id, "findShift")}
                                                    </div>
                                                }
                                            })}
                                    </div>
                                </>
                            )
                        })
                    : null
                    }
                </div>
            </div>
        </div>
    )
}