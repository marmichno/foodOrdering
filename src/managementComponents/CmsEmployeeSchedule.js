import {CmsNavbar} from './CmsNavbar';
import {useState, useEffect} from 'react';

export const CmsEmployeesSchedule = () =>{

    const newDate = new Date();

    const [selectedMonth, setSelectedMonth] = useState(newDate.getMonth());
    const [currentYear, setCurrentYear] = useState(newDate.getYear());

    const monthNames = ["January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December"];

    const employees = ['filler', 'employee1', 'employee1', 'employee1', 'employee1', 'employee1', 'employee1', 'employee1', 'employee1'];

    let month = newDate.getMonth();



    const createArrayOfDays = (selectedMonth, currentYear) => {
        let arrayOfDays = [];

        let year = newDate.getYear();
        let daysInMonth = new Date(currentYear, selectedMonth + 1, 0).getDate();

        console.log(daysInMonth);

        const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 
        'Thursday', 'Friday', 'Saturday'];

        for(let i = 0; i < daysInMonth; i++){
            let whichDayIsIt = new Date(currentYear, selectedMonth + 1, i).getDay();
            arrayOfDays.push(
                {
                    dayDate: i + 1,
                    dayName: days[whichDayIsIt]
                }
            );
        }

        return arrayOfDays;
    }

    
    const showDate = () =>{
        // console.log(year, monthNames[month], date);
        console.log(createArrayOfDays(selectedMonth, currentYear));
    }

    return(
        <div className="cmsEmployeesScheduleContainer">
            <CmsNavbar/>
            <div className="cmsEmployeesScheduleContent">
                <div className="shiftsContainer">
                    <div className="shiftsInfoContainer">
                        <div className="shiftInfo"><h1>1 - 06:00 - 14:00</h1></div>
                        <div className="shiftInfo"><h1>2 - 07:00 - 15:00</h1></div>
                        <div className="shiftInfo"><h1>3 - 08:30 - 16:30</h1></div>
                        <div className="shiftInfo"><h1>4 - 13:40 - 21:40</h1></div>
                        <div className="shiftInfo"><h1>5 - 15:00 - 23:00</h1></div>
                        <div className="shiftInfo"><h1>6 - 16:30 - 00:30</h1></div>
                        <div className="shiftInfo"><h1>6 - 16:30 - 00:30</h1></div>
                        <div className="shiftInfo"><h1>6 - 16:30 - 00:30</h1></div>
                        <div className="shiftInfo"><h1>6 - 16:30 - 00:30</h1></div>
                        <div className="shiftInfo"><h1>6 - 16:30 - 00:30</h1></div>
                        <div className="shiftInfo"><h1>6 - 16:30 - 00:30</h1></div>
                    </div>
                    <div className="addShifts">
                        <input type="text" placeholder="hour range ex: 16:30-24:30" autocomplete="off" name="category" required></input>
                        <input className="shiftId" type="number" placeholder="shift id ex: 1" autocomplete="off" name="category" required></input>
                        <button>add new shift</button>
                    </div>
                </div>
                <div className="calendarContainer">
                    {employees.map(employee => {
                        return(
                            <>
                                <div className="employeeRow">
                                    <div className="employee">{employee}</div>
                                        {createArrayOfDays(selectedMonth, currentYear).map(value => {
                                            if(employee === 'filler'){
                                                return( 
                                                    <div>{value.dayDate}</div>
                                                )
                                            }
                                            if(value.dayName === 'Saturday' || value.dayName === 'Sunday'){
                                                return <div style={{backgroundColor: 'orange'}}></div>
                                            }else{
                                                return <div></div>
                                            }
                                        })}
                                </div>
                            </>
                        )
                    })}
                </div>
            </div>
        </div>
    )
}