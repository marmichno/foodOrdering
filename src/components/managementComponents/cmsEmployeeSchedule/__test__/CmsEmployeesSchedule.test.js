jest.unmock('axios');
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import {CmsEmployeesSchedule} from '../CmsEmployeeSchedule';
import "@testing-library/jest-dom/extend-expect";
import {render, fireEvent, waitFor, screen} from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux'; 
import validEmployeeJson from '../__mocks__/validEmployee.json';
import validSchedulesJson from '../__mocks__/validSchedules.json';
import validShiftsJson from '../__mocks__/validShifts.json';
import TestRenderer from "react-test-renderer";

const store = createStore(() => [], {}, applyMiddleware());

const mock = new MockAdapter(axios);

test("checks if shifts are visible after api call", async () => {

    mock.onGet("http://localhost:8080/api/employee").reply(200, validEmployeeJson);
    mock.onGet("http://localhost:8080/api/schedule/2021/6").reply(200, validSchedulesJson);
    mock.onGet("http://localhost:8080/api/shift").reply(200, validShiftsJson);

    const {getAllByText} = render(<Provider store={store}><CmsEmployeesSchedule/></Provider>);

    await waitFor(() => {
        expect(getAllByText('day off'));
        expect(getAllByText('513 - 16:30-24:30'));
        expect(getAllByText('547 - 16:30-24:30'));
        expect(getAllByText('546 - 16:30-24:30'));
        expect(getAllByText('548 - 16:30-24:30'));
    })
});

test("checks if employees are visible after api call", async () => {

    mock.onGet("http://localhost:8080/api/employee").reply(200, validEmployeeJson);
    mock.onGet("http://localhost:8080/api/schedule/2021/6").reply(200, validSchedulesJson);
    mock.onGet("http://localhost:8080/api/shift").reply(200, validShiftsJson);

    const {getAllByText} = render(<Provider store={store}><CmsEmployeesSchedule/></Provider>);

    await waitFor(() => {
        expect(getAllByText('Marcin'));
    })
});

test("checks if shift could be added to schedule", async () => {

    mock.onGet("http://localhost:8080/api/employee").reply(200, validEmployeeJson);
    mock.onGet("http://localhost:8080/api/schedule/2021/6").reply(200, validSchedulesJson);
    mock.onGet("http://localhost:8080/api/shift").reply(200, validShiftsJson);
    mock.onPost("http://localhost:8080/api/schedule").reply(200);

    const {getAllByText, getByTestId} = render(<Provider store={store}><CmsEmployeesSchedule/></Provider>);

    await waitFor(() => {
        fireEvent.click(getByTestId("shift1"));
    });

    await waitFor(() => {
        fireEvent.click(getByTestId("numOfDay9_482"));
    });

    await waitFor(() => {
        expect(mock.history.post[0].data).toBe(JSON.stringify({"date":"2021-06-10","shiftEntity":{"id":546},"employeeEntity":{"id":482}}));
    })
    
    mock.resetHistory();
});

test("checks if shift could be deleted from schedule", async () => {

    mock.onGet("http://localhost:8080/api/employee").reply(200, validEmployeeJson);
    mock.onGet("http://localhost:8080/api/schedule/2021/6").reply(200, validSchedulesJson);
    mock.onGet("http://localhost:8080/api/shift").reply(200, validShiftsJson);
    mock.onPost("http://localhost:8080/api/schedule").reply(200);
    mock.onDelete("http://localhost:8080/api/schedule/609").reply(200);

    const {getAllByText, getByTestId} = render(<Provider store={store}><CmsEmployeesSchedule/></Provider>);

    await waitFor(() => {
        fireEvent.click(getByTestId("dayOff"));
    })

    await waitFor(() => {
        fireEvent.click(getByTestId("numOfDay0_482"));
    });

    await waitFor(() => {
        expect(mock.history.delete[0].url).toBe('http://localhost:8080/api/schedule/609');
    });
});


test("matches snapshot", () => {
    const tree = TestRenderer.create(<Provider store={store}><CmsEmployeesSchedule/></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
});
