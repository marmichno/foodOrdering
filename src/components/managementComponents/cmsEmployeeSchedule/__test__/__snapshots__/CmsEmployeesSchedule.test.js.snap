// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="cmsEmployeesScheduleContainer"
>
  <div
    className="hamburgerMenu"
  >
    <svg
      fill="none"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="3"
        x2="21"
        y1="12"
        y2="12"
      />
      <line
        x1="3"
        x2="21"
        y1="6"
        y2="6"
      />
      <line
        x1="3"
        x2="21"
        y1="18"
        y2="18"
      />
    </svg>
  </div>
  <div
    className="cmsEmployeesScheduleContent"
  >
    <div
      className="shiftsContainer"
    >
      <div
        className="shiftsInfoContainer"
      >
        <div
          className="shiftInfo"
        >
          <input
            autoComplete="off"
            name="category"
            onChange={[Function]}
            placeholder="shift id"
            required={true}
            type="number"
          />
          <button
            onClick={[Function]}
          >
            delete
          </button>
        </div>
        <div
          className="shiftInfo"
        >
          <h1
            data-shift={0}
            data-testid="dayOff"
            onClick={[Function]}
          >
            day off
          </h1>
        </div>
      </div>
      <div
        className="addShifts"
      >
        <input
          autoComplete="off"
          name="category"
          onChange={[Function]}
          placeholder="hour range ex: 16:30-24:30"
          required={true}
          type="text"
        />
        <button
          onClick={[Function]}
        >
          add new shift
        </button>
      </div>
    </div>
    <div
      className="calendarContainer"
    >
      <div
        className="changeDate"
      >
        <button
          onClick={[Function]}
        >
          previous
        </button>
        <h1>
          2021, June
        </h1>
        <button
          onClick={[Function]}
        >
          next
        </button>
      </div>
      <div
        className="employeeRow"
      >
        <div
          className="employee"
        />
        <div>
          <b>
            01
          </b>
        </div>
        <div>
          <b>
            02
          </b>
        </div>
        <div>
          <b>
            03
          </b>
        </div>
        <div>
          <b>
            04
          </b>
        </div>
        <div>
          <b>
            05
          </b>
        </div>
        <div>
          <b>
            06
          </b>
        </div>
        <div>
          <b>
            07
          </b>
        </div>
        <div>
          <b>
            08
          </b>
        </div>
        <div>
          <b>
            09
          </b>
        </div>
        <div>
          <b>
            10
          </b>
        </div>
        <div>
          <b>
            11
          </b>
        </div>
        <div>
          <b>
            12
          </b>
        </div>
        <div>
          <b>
            13
          </b>
        </div>
        <div>
          <b>
            14
          </b>
        </div>
        <div>
          <b>
            15
          </b>
        </div>
        <div>
          <b>
            16
          </b>
        </div>
        <div>
          <b>
            17
          </b>
        </div>
        <div>
          <b>
            18
          </b>
        </div>
        <div>
          <b>
            19
          </b>
        </div>
        <div>
          <b>
            20
          </b>
        </div>
        <div>
          <b>
            21
          </b>
        </div>
        <div>
          <b>
            22
          </b>
        </div>
        <div>
          <b>
            23
          </b>
        </div>
        <div>
          <b>
            24
          </b>
        </div>
        <div>
          <b>
            25
          </b>
        </div>
        <div>
          <b>
            26
          </b>
        </div>
        <div>
          <b>
            27
          </b>
        </div>
        <div>
          <b>
            28
          </b>
        </div>
        <div>
          <b>
            29
          </b>
        </div>
        <div>
          <b>
            30
          </b>
        </div>
        <div>
          <b>
            31
          </b>
        </div>
      </div>
    </div>
  </div>
</div>
`;
